packages/core/src/tracing/instrumentation/database/db2.js

```diff
@@ -0,0 +1,539 @@
+        finishSpan(ctx, null, span);
+        return err;
+      }
+    }
+
+    const customerCallbackIndex =
+      // eslint-disable-next-line no-nested-ternary
+      originalArgs.length === 2 && typeof originalArgs[1] === 'function'
+        ? 1
+        : originalArgs.length === 3 && typeof originalArgs[2] === 'function'
+        ? 2
+        : null;
+    const customerCallback = originalArgs[customerCallbackIndex];
+
+    if (customerCallback) {
+      originalArgs[customerCallbackIndex] = function instanaCallback(err) {
+        if (err) {
+          span.ec = 1;
+          span.data.db2.error = tracingUtil.getErrorDetails(err);
+        }
+
+        const result = customerCallback.apply(this, arguments);
+        finishSpan(ctx, result, span);
+        return result;
+      };
+
+      return originalFunction.apply(ctx, originalArgs);
+    }
+  });
+}
+
+function createSpan(stmt, fn) {
+  // eslint-disable-next-line max-len
+  // https://github.ibm.com/instana/backend/blob/develop/forge/src/main/java/com/instana/forge/connection/database/ibmdb2/IbmDb2Span.java
+  const span = cls.startSpan(exports.spanName, constants.EXIT);
+  span.stack = tracingUtil.getStackTrace(fn);
+  span.d = Date.now() - span.ts;
+  span.data.db2 = {
+    stmt: tracingUtil.shortenDatabaseStatement(stmt),
+    dsn: tracingUtil.sanitizeConnectionStr(connectionStr)
+  };
+
+  return span;
+}
+
+function finishSpan(ctx, result, span) {
+  captureFetchError(result, span);
+
+  if (ctx.conn.inTransaction) {
+    return handleTransaction(ctx, span);
+  }
+
+  // NOTE: This signalises us the end of the trace
+  //       Because we want to capture errors till the customer finished the operation.
+  //       That's why we want to wait for this call. The ibm docs say, it's required to call it!
+  if (result && result.closeSync) {
+    const originalCloseSync = result.closeSync;
+    let closeSyncCalled = false;
+    result.closeSync = function instanaCloseSync() {
+      if (closeSyncCalled) return originalCloseSync.apply(this, arguments);
+
+      closeSyncCalled = true;
+      span.transmit();
+      return originalCloseSync.apply(this, arguments);
+    };
+
+    // CASE: customer forgets to call `result.closeSync`
+    //       search for result.closeSync() in:
+    //       https://github.com/ibmdb/node-ibm_db/blob/master/APIDocumentation.md
+    // TODO: discuss timeout length and the approach in general
+    setTimeout(() => {
+      if (!closeSyncCalled) {
+        closeSyncCalled = true;
+        span.transmit();
+      }
+    }, 500);
+  } else {
+    span.transmit();
+  }
+}
+
+function handleTransaction(ctx, span) {
+  const originalEndTransaction = ctx.conn.endTransaction;
+  const originalEndTransactionSync = ctx.conn.endTransactionSync;
+
+  // NOTE: This is an internal fn to avoid instrumenting commit, rollback separately
+  ctx.conn.endTransaction = function instanaEndTransaction(rollback) {
+    if (rollback) {
+      span.cancel();
```

Comment:
_question:_ Should we cancel a span if a rollback happens or should that rather result in a span with an error? I tend to think the latter. The interaction with the database has actually happened, so I think this should be a span. We should check what we do for other db instrumentations.

Reply:
Yeah true, let's trace a rollback 👍

Reply:
Done 👍